cli/src/cli_util.rs

```diff
@@ -2586,6 +2587,57 @@ pub struct EarlyArgs {
     pub config_toml: Vec<String>,
 }
 
+#[derive(clap::Args, Clone, Debug)]
+pub struct ShowAllocStats {
+    /// Show memory allocation statistics from the internal heap allocator
+    /// on `stdout`, when the program exits.
+    #[arg(long, global = true)]
+    show_heap_stats: bool,
+}
+
+/// Lazy global static. Used only to defer printing mimalloc stats until the
+/// program exits, if set to `true`.
+static PRINT_HEAP_STATS: OnceLock<bool> = OnceLock::new();
+
+/// Enable heap statistics for the user interface; should be used with
+/// [`CliRunner::add_global_args`]. Does nothing if the memory allocator is
+/// unused, i.e. `#[global_allocator]` is not set to mimalloc in your program.
+pub fn heap_stats_enable(_ui: &mut Ui, opts: ShowAllocStats) -> Result<(), CommandError> {
+    if opts.show_heap_stats {
+        PRINT_HEAP_STATS.set(true).unwrap();
+    }
+    Ok(())
+}
+
+/// Reset heap allocation statistics for the memory allocator. Often used at the
+/// very beginning of the program (to clear allocations that may happen before
+/// `_main` execution.)
+pub fn heap_stats_reset() {
+    mimalloc::stats_reset();
+}
+
+/// Print heap allocation statistics to `stderr`, if enabled.
+pub fn heap_stats_print() {
+    // NOTE (aseipp): can we do our own custom printing here? it's kind of ugly
+    if PRINT_HEAP_STATS.get() == Some(&true) {
+        eprintln!("========================================");
+        eprintln!("mimalloc memory allocation statistics:\n");
+        mimalloc::stats_print(&|l| {
+            eprint!("{}", l.to_string_lossy());
+        });
+    }
+}
+
+/// Wrap a given closure with calls to [`heap_stats_reset`] and
+/// [`heap_stats_print`], and return the result of the closure. Useful for
+/// conveiently printing heap allocation statistics for a given function body.
+pub fn heap_stats_with_closure<R>(f: impl FnOnce() -> R) -> R {
+    heap_stats_reset();
+    let result = f();
+    heap_stats_print();
+    result
+}
+
```

Comment:
In my meager opinion, this integration belongs into the actual `MiMalloc` crate as it's dead code when Mimalloc is not enabled.

Reply:
I've moved all of this into `main.rs`, since I believe that that's the right place for it; both outside the `jj-cli` crate and also `jj-cbits`. Users who want to rely on `jj-cli` (like Google) will have to write their own `main.rs` anyway, so they can choose what they want there.